### PR TITLE
WIP: Adding cache to Scapegoat & Scalastyle Inspections Generator Tasks

### PR DESCRIPTION
**Pull request by BalmungSan**
Ok so, being honest I am doing this more like a PoC, because I think I will need this in my project at the office. Since it does not bring too much value to this project.

Nevertheless I rather doing it here to share it, than keeping it to myself.
Hope you _(and anybody checking the repo in the future)_ will find it useful 😉 

For now I had only done it for the **Scapegoat** _task_.
I just tested it, and it works as intended: On consequent calls to `compile`  it does not re-run the task :D

I think I can easily replicate the work for the **Scalastyle** _task_.
However, I do not know if it will work as intended, since the template file is part of the main sources. Then, a change on such file should only had effect after a `reload` of **SBT**.

So this serves as the excuse to bring up and old discussion related with the generator tasks.
I think that the template files should be on the resources folder, instead of on the scala folder. That way, we can mark all, but the `ScalastyleInspections` object, as `private[scalastyle]`.
Although, having them in the source, allows to write more typesafe tests, instead of the _"String checks"_ I had for the `ScapegoatInspectionsGenerator`.

Thus, I do not know if there is a way _(maybe using scala-meta)_ to keep the templates as resources. But, having typesafe tests.
Or, since the generated files become part of the main project, we may move those tests to the main project. And in the meta-project, just keep a couple of simple tests to verify the generation logic.

### CODE REVIEW

**Comment by mwz** [comment]
That's nice 👍

> However, I do not know if it will work as intended, since the template file is part of the main sources. Then, a change on such file should only had effect after a `reload` of **SBT**.

I'm not sure how this would work, but I suppose if it does work that way it should be ok because we don't change those files too often for this to become a pain.

> So this serves as the excuse to bring up and old discussion related with the generator tasks.
> I think that the template files should be on the resources folder, instead of on the scala folder.

Personally, I would have those files as sources, because technically even though those are templates (and you might want to consider them as resources), we still want them to be valid source files that can be compiled - just with missing implementation. When I want to change those files I open the meta-project just like any other project in my IDE and I can work directly with the source files, otherwise, it's quite difficult to work with those without the full IDE support.

**Comment by BalmungSan** [comment]
First of all, sorry for taking so long to answer. As I said before, I will be really busy for a while because I started a new project.

> "Personally, I would have those files as sources, because technically even though those are templates (and you might want to consider them as resources), we still want them to be valid source files that can be compiled - just with missing implementation".

Well, anyways after being filled they will be compiled and tested on the project scope.
Thus, I really do not see any value on doing that on the meta-project scope too.

> "When I want to change those files I open the meta-project just like any other project in my IDE and I can work directly with the source files, otherwise, it's quite difficult to work with those without the full IDE support".

You may have a point there...
However, I really used to work with plain vim / emacs. So, loosing IDE _"capabilities"_ had never bother me. And these files are pretty small / simple so not big deal.
But that is just me 😅 I wont use that as an argument.

In conclusion.
I still believe it would be better to treat the templates as _resources_ and move the tests to the main project.
But, if you still want to leave the **ScalastyleInspectionsTemplate** as a source, I will just add the cache to the generator.

**Comment by mwz** [comment]
> And these files are pretty small / simple so not big deal.

yeah, ok - I don't feel too strong about this to block your PR, so feel free to move those files into resources :smile:

**Comment by BalmungSan** [comment]
Ok so, I just tested that all the examples projects just worked as intended with the changes 👌 

Also, now after the first compilations the generated files wont be generated anymore unless you clean the `target` folder or if the templates gets modified.
If you modify the generators itself you need to execute a `reload` in order to the changes take effect.

I had to admit that having the tests of the generators as plain strings comparisons are a pain in the ass, but I guess that is not so bad, given we could make all these classes privates to their packages.

Anyways, I want to know what do you think of all this.

File: project/src/test/scala/ScalastyleInspectionsGeneratorTest.scala [hunk 1]
```diff
@@ -32,155 +32,152 @@ class ScalastyleInspectionsGeneratorTest extends FlatSpec with Matchers {
...
+        |  params = List(
+        |    Param(
+        |      name = "maxFileLength",
+        |      typ = IntegerType,
+        |      label = "Maximum file length",
+        |      description = "Maximum number of lines in a file",
+        |      default = "1500"
+        |    )
+        |  )
+        |)""".stripMargin
```
Comment (mwz) [inline project/src/test/scala/ScalastyleInspectionsGeneratorTest.scala]: Same here, making this a string really feels unnecessary to me.

File: src/test/scala/com/mwz/sonar/scala/scalastyle/ScalastyleRulesRepositorySpec.scala [hunk 2]
```diff
@@ -48,7 +48,7 @@ class ScalastyleRulesRepositorySpec extends FlatSpec with Matchers with Inspecto
 
   it should "include all Scalastyle inspections" in new Ctx {
     ScalastyleInspections.AllInspections should have size 69 // 29 templates + 40 default rules
-    ScalastyleInspections.AllInspectionsByClass.size shouldBe ScalastyleInspections.AllInspections.size
+    ScalastyleInspections.AllInspectionsByClass.size shouldBe 69
```
Comment (mwz) [inline src/test/scala/com/mwz/sonar/scala/scalastyle/ScalastyleRulesRepositorySpec.scala]: 👍

File: project/src/main/resources/ScapegoatInspections.scala [hunk 3]
```diff
@@ -20,7 +20,7 @@ package scapegoat
 
 import org.sonar.api.batch.rule.Severity
 
-private[scapegoat] sealed trait Level {
+private[scapegoat] sealed trait Level extends Product with Serializable {
```
[the remark this answers is not in the document]
Comment (BalmungSan) [reply project/src/main/resources/ScapegoatInspections.scala]: https://typelevel.org/blog/2018/05/09/product-with-serializable.html

TL;DR; All case classes / objects  `extend Product with Serializable`, thus it is a good idea to make the root trait to extend it, that help a little bit to type inference, specially when using collections with **ADTs** .

On our case we do not need it, but now i am just used to do it when defining **ADTs**, is like a personal best practice. However, If you want, I can remove it since I do not have a strong argument for this case.

File: project/src/main/scala/ScalastyleInspectionsGenerator.scala [hunk 4]
```diff
@@ -31,41 +29,50 @@ import scala.xml.{Node, NodeSeq, XML}
...
-    val documentationStream: InputStream = c.getResourceAsStream("/scalastyle_documentation.xml")
-    val inspections: NodeSeq = XML.load(definitionStream) \\ "checker"
-    val docs: NodeSeq = XML.load(documentationStream) \\ "scalastyle-documentation" \ "check"
-    val config: Config = ConfigFactory.load(c.getClassLoader)
+          val c = classOf[MainConfig]
+          val definitionStream = c.getResourceAsStream("/scalastyle_definition.xml")
+          val documentationStream = c.getResourceAsStream("/scalastyle_documentation.xml")
+          val inspections = XML.load(definitionStream) \\ "checker"
+          val docs = XML.load(documentationStream) \\ "scalastyle-documentation" \ "check"
+          val config = ConfigFactory.load(c.getClassLoader)
```
[the remark this answers is not in the document]
Comment (BalmungSan) [reply project/src/main/scala/ScalastyleInspectionsGenerator.scala]: Uhm... ok.
I really love type inference, but I guess it makes the code more readable.

**Comment by BalmungSan** [comment]
Ok so for addressing all other comments, the thing is that since I moved the **ScalastyleInspections.scala** file to resources, it is not longer compiled. Thus, on the _meta-project_ `ScalastyleInspection` no longer exists.

That was the real trade-off of moving the file to the resources.
For example, if you see the **Scapegoat** generator and its tests everything is a String.
When we discussed this the last time, that was the point.
I wanted to move them to resources to have them private, but the problem was that the code and the tests becomes very _"Stringly typed"_ _(Using string instead of proper types - you asked for the definition the last time)_.

Nevertheless, I really do not like having those Strings, testing them becomes a real problem.
But, since the generators probably will not get changed constantly, maybe is worth having it.
In any case, the generated files will be compiled and tested on the main. So, if the generation logic is bad, we will notice it. Maybe those tests are not necessary at all?

Being honest, I was very surprised when you said

> "yeah, ok - I don't feel too strong about this to block your PR, so feel free to move those files into resources ".

Last time you really did not want to change that, and it seems this time either, only that you forgot my evil intentions 😂 

-----

So, it is your decision, I can revert the changes.
However, again, if the template is not a resource, it does not make sense to cache it, especially because the meta-project should be recompiled first... _(anyways, that does not happen even without cache, so is easy to break everything)_.
But, again, I can add the cache even with that if you want.
Or, we could just close this, it does not bring too much value at all _(specially if we only add cache to **Scapegoat** generator)_. As I said when opening the PR, I did this mainly to learn because I need it for a project at the office.

File: project/src/main/scala/ScalastyleInspectionsGenerator.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (mwz) [reply project/src/main/scala/ScalastyleInspectionsGenerator.scala]: Yes, type inference is great, but it's definitely easier to reason about the code knowing what the types are.

File: project/src/main/resources/ScapegoatInspections.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mwz) [reply project/src/main/resources/ScapegoatInspections.scala]: Tbh, I've never seen before anyone else do it and I think the problem goes away when you explicitly give types to your values, which is what I tend to do and that's probably why I've never run into this issue. But let's keep it if that's your preference.

**Comment by mwz** [comment]
> Being honest, I was very surprised when you said
> 
> > "yeah, ok - I don't feel too strong about this to block your PR, so feel free to move those files into resources ".
> 
> Last time you really did not want to change that, and it seems this time either, only that you forgot my evil intentions 😂

Sorry, I don't think that I fully understood the implications of those changes, especially the "stringification" aspect of it. Seeing those changes now, I think that the downsides outweigh the benefits, especially having in mind the improvement of generating code tree in a more generic way, turning the generated code into a string doesn't feel to me like a step in the right direction. I think that trading an extra one or maybe two seconds of compile time for improved maintainability isn't a bad trade to make 😄 

> But, since the generators probably will not get changed constantly, maybe is worth having it.

I'm planning to make some changes as a part of https://github.com/sksamuel/scapegoat/issues/219, so it would be a massive pain having to change those strings everywhere.

**Comment by BalmungSan** [comment]
Ok, fair enough. I will close the PR but will leave the branch alive in case I / You / or anyone else want to take a look to a cached generator example.
Probably I will make a **gist** too.

_(My evil intentions blast off again! 😆)_

**Comment by mwz** [comment]
Haha, cool thanks for the effort. I hope you will find caching useful at work!
